### ISSUE TITLE
Request: use aspects the way DEBUG uses them

### ISSUE THREAD

**Issue by fernandocanizo** [woveon/woveon-logger#4]
I came too excited to try `woveon` logger attracted by the _aspect_ feature, only to find out it works on a line-by-line basis.

My expectation was that this logger had a mix between levels plus the ability to turn on/off logging for those already written messages.

I'm wondering if you would give this a second thought.

Ideally the ability to define aspects and turn them on/off is to only see part of the logging, so you can better debug the code. On big code bases, there's usually too much logging going on, so when a bug appears, one wants to turn off some parts, or maybe turn on only a specific part and ignore the rest.

The way `woveon` logger works forces you to write those specific lines, instead of using the already written ones.

Maybe my explanation got over-complicated. Let me express it with code:

```
const woveonLogger = require('woveon-logger');
const log = new woveonLogger({ aspect: 'moduleX' });

const foo = () => log.info('some useful info from foo()');

module.exports = foo;
```

Then by running the app with `WOV_LOGGER_ASPECTS=moduleX node server.js` we get only the log lines belonging to moduleX.

Ideally a string like the one used by [debug](https://github.com/visionmedia/debug) would be great, so one can say `WOV_LOGGER_ASPECTS=app:*` to log everything, or `WOV_LOGGER_ASPECTS=app:moduleX` to log only moduleX, or `WOV_LOGGER_ASPECTS=app:*,-moduleX` to log everything except moduleX.

Would you say it's a possible direction for `woveon` logger to take or you're not interested in such functionality?

**Comment by cwingrav** [comment]
Howdy Fernando. Thanks for the feedback. 

I use aspects constantly, to enabled/disable logging of parts of my file, and to configure it at program start (via the `WOV_LOGGER_ASPECTS` env variable) but also at runtime via `setAspect(_string)`. The environment variable can also toggle multiple aspects, even turning them off with '!'. 

I believe that is what your are asking about? 

so: `node server.js`

```
// server.js
const logger = new (require('woveon-logger'))();

function hello() {
  logger.log('hello');
  logger.aspect('A', 'foo');
  logger.aspect('B', 'bar');
}

hello(); // logs nothing
logger.setAspect('A');
hello(); // logs 'foo'
logger.setAspect('B');
hello(); // logs 'foo' and 'bar'
logger.setAspect('A', false);
hello(); // logs 'bar'
```

Outputs:

```
[2018-08-06T22:58:29.046Z] [INFO--] hello
[2018-08-06T22:58:29.048Z] [INFO--] hello
[2018-08-06T22:58:29.048Z] [A-----] foo
[2018-08-06T22:58:29.049Z] [INFO--] hello
[2018-08-06T22:58:29.049Z] [A-----] foo
[2018-08-06T22:58:29.049Z] [B-----] bar
[2018-08-06T22:58:29.049Z] [INFO--] hello
[2018-08-06T22:58:29.049Z] [B-----] bar
```

Or, using environment variables: `WOV_LOGGER_ASPECTS='A B' node server.js` 

Outputs:

```
[2018-08-06T22:58:52.771Z] [INFO--] hello
[2018-08-06T22:58:52.773Z] [A-----] foo
[2018-08-06T22:58:52.773Z] [B-----] bar
[2018-08-06T22:58:52.773Z] [INFO--] hello
[2018-08-06T22:58:52.774Z] [A-----] foo
[2018-08-06T22:58:52.774Z] [B-----] bar
[2018-08-06T22:58:52.774Z] [INFO--] hello
[2018-08-06T22:58:52.774Z] [A-----] foo
[2018-08-06T22:58:52.774Z] [B-----] bar
[2018-08-06T22:58:52.774Z] [INFO--] hello
[2018-08-06T22:58:52.774Z] [B-----] bar
```

Or, using environment variables, turning off A: `WOV_LOGGER_ASPECTS='A B !A' node server.js`

Outputs:

```
[2018-08-06T23:02:11.126Z] [INFO--] hello
[2018-08-06T23:02:11.128Z] [B-----] bar
[2018-08-06T23:02:11.128Z] [INFO--] hello
[2018-08-06T23:02:11.129Z] [A-----] foo
[2018-08-06T23:02:11.129Z] [B-----] bar
[2018-08-06T23:02:11.129Z] [INFO--] hello
[2018-08-06T23:02:11.129Z] [A-----] foo
[2018-08-06T23:02:11.129Z] [B-----] bar
[2018-08-06T23:02:11.129Z] [INFO--] hello
[2018-08-06T23:02:11.129Z] [B-----] bar
```

**Comment by fernandocanizo** [comment]
Nope, probably I entangled myself on the explanation. Let me try again:

Let's assume an Express server which calls both `module01.js` and `module02.js` to answer to some request, then, what I'd want is the ability to debug only one of them, **without prior aspect setting on a line-by-line basis, but a module basis**. I'll write it _ala debug_ style:

```
// module01.js
// in debug syntax this would be:
// require('debug')('myapp:module01')
// a possible syntax implementation which resembles debug:
log = require('woveon')({ aspect: 'myapp:module01' });

module.exports = () => log.info('Logging from module01');

// module02.js
log = require('woveon')({ aspect: 'myapp:module02' });

module.exports = () => log.info('Useful info from m02');
```

Having that in place now you can call `DEBUG=myapp:*` and get:

```
[2018-08-06T23:02:11.126Z] [INFO--] Logging from module01
[2018-08-06T23:02:11.128Z] [INFO--] Useful info from m02
```

With `DEBUG=myapp:module01`:

```
[2018-08-06T23:02:11.126Z] [INFO--] Logging from module01
```

With `DEBUG=myapp:module02`:

```
[2018-08-06T23:02:11.128Z] [INFO--] Useful info from m02
```

Omitting one: `DEBUG=myapp:*,-myapp:module01`:

```
[2018-08-06T23:02:11.126Z] [INFO--] log from other place
[2018-08-06T23:02:11.126Z] [INFO--] and another one
[2018-08-06T23:02:11.126Z] [INFO--] and yet some other place not being m01 nor m02
[2018-08-06T23:02:11.128Z] [INFO--] Useful info from m02
```

I used `DEBUG` variable but you can use `WOV_*` or honor `DEBUG`.

I found your package after searching for a logging solution that would allow me to have levels **and** aspects _ala debug_. The former to define logging levels that will allow me to activate more verbose ones as I debug/inspect some code. The later to allow me to turn on/off specific parts of the logging on my code.

The way you have currently defined your aspects doesn't allow me to change what's being logged on the fly without having written those aspect-oriented lines in my code previously.

The way I propose to use aspects doesn't require one to set aspect logging on each possible  part of the code you want to inspect, but only at a module level.

Here the word "module" doesn't necessarily refers to a file, a module can span several files. Maybe we should use the word "section" o "area" or something broader.

Anyway, what I'm requesting is nothing different than what [debug](https://github.com/visionmedia/debug) does, so if you don't know it, go take a look at it.

I hope I explained myself better this time, and thanks for taking the time to read me.

**Comment by cwingrav** [comment]
See #5 regarding the regular expressions. 

If you like what I said about a global logger, I can probably find some time for that. I can see myself using that in my own projects.

**Comment by fernandocanizo** [comment]
Thanks for your time Chad, I really appreciate you take the time to answer me, however it seems we're not on the same page regarding our logging needs.

The main difference is you use `logger.aspect()` explicitly, while what I want is to set it at the beginning and manage all levels from the environment.

I don't want any calls inside my code to turn on/off the aspect. I just want [DEBUG](https://github.com/visionmedia/debug#readme)-like on/off ability + levels. Something like this pseudo-code:

```
const woveon = require('woveon-logger');
const log = new woveon({ aspect: 'sectionX' });

log.info('some info...');
log.warning('some warning...');
log.error('some error...');
// ...
```

Then I can turn on by calling it with `WOV_LOGGER_ASPECTS='sectionX'` or off with `WOV_LOGGER_ASPECTS='-sectionX'`.

I think changing colors dynamically is convoluted and confusing, so I think an aspect should only manage the visible status of a log line, not it's color or any other property. But that's me.

Anyway, thanks again for your time :)

**Comment by cwingrav** [comment]
Well, I'm always appreciating the feedback. When I set my loggers, I generally set them and forget them, usually in a make file, at the start. Just like what you want. 

It's a work in progress. 

Also, the code I provided last time was pseudocode, to achieve some of the effects you wanted. It wasn't meant to run. I created an issue to implement some of those features.

Cheers and hope you give it a try in some future project.
